### ISSUE TITLE
Proposal: provide events before and after GC

### ISSUE THREAD

**Issue by ygc369** [dotnet/coreclr#16931]
Sometimes I want to do something before and/or after GC, but there's no way to do it now.
That's very disappointing, please trigger events before and after GC, thus developers could subscribe them and do what they want to do in event handlers once these events happen. The event arguments should include GC generation at least.

What people want to do before/after GC is usually something about object caching pool. 

Situation 1: Use weak reference list to cache some objects. 
In this situation, the targets of many weak references will become null after GC, so people may want to pick these weak references out of the list after GC.

Situation 2: Use strong reference list to cache some objects.
In this situation, people may want to assign some strong references to be null just before GC, thus their memory space may be collected by GC soon. People can reuse cached objects until GC happens, and need not worry about memory leak if they don't reuse later.

**Comment by ygc369** [comment]
Nobody interested in it?

**Comment by jkotas** [comment]
- The event handlers would need to be written in a very special way to avoid allocations.
- The events would serialize execution. It would have negative performance consequences for parallel GC.

GC handles that can be used to achieve equivalent effect avoids both these problems.

**Comment by ygc369** [comment]
@jkotas 
>GC handles that can be used to achieve equivalent effect avoids both these problems.

How to? Could you give an example?

**Comment by jkotas** [comment]
> the targets of many weak references will become null after GC, so people may want to pick these weak references out of the list after GC.

This can be done by allocating dummy finalizable object. The finalizer will run short after GC collects the object. The finalizer is a regular code that does not block GC.

> people may want to assign some strong references to be null just before GC, thus their memory space may be collected by GC soon

This sounds like some kind of caching GC handle: Handle that is not completely weak, but that the GC can still set to null based on the memory pressure. GC handle like this does not exist as first class concept today.

**Comment by ygc369** [comment]
@jkotas 

> This sounds like some kind of caching GC handle: Handle that is not completely weak, but that the GC can still set to null based on the memory pressure. GC handle like this does not exist as first class concept today.

Yes, I want this kind of caching GC handle, does the CLR team have any plan to provide it?

**Comment by jkotas** [comment]
We are researching how to make GC work better with caching. We do not have a firm plan yet.
cc @Maoni0 @davmason

**Comment by Maoni0** [comment]
>Sometimes I want to do something before and/or after GC, but there's no way to do it now.

there are multiple ways to do things before and after a GC. you can do this via profiling

[ICorProfilerCallback2::GarbageCollectionStarted](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilercallback2-garbagecollectionstarted-method)
[ICorProfilerCallback2::GarbageCollectionFinished](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilercallback2-garbagecollectionfinished-method)

or hosting APIs

[IGCThreadControl](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/hosting/igcthreadcontrol-interface)

**Comment by Maoni0** [comment]
yeah, that was intentional because a GC can happen while you are doing things in managed code. so doing things "before a GC" doesn't mean a whole lot in managed code because another GC (or GCs) could happen while those things are happening.

**Comment by ygc369** [comment]
@jkotas @Maoni0 
It seems that .Net Core 2.2 has runtime events.(see https://blogs.msdn.microsoft.com/dotnet/2018/12/04/announcing-net-core-2-2/)
Could that solve this problem?

**Comment by jkotas** [comment]
The runtime events are asynchronous. You will receive them some time after they actually happened.

**Comment by ygc369** [comment]
@jkotas 
I need synchronous callback before GC, which guarantees that GC would not start to work before the callback returns. I know this might impact performance of GC. But if nobody registers such callback, there would be no impact to GC. If somebody registers such callback, he must be responsible for the impact, that means he is sure that the impact is nothing serious in his use case/scenario. Of course, the synchronous callback before GC must be short and fast, and can not make any heap allocation.
